core/mapi/cache/namedPropCache.cpp

```diff
@@ -31,88 +34,179 @@ namespace cache
 					mapi::setTag(lpTag, i) = PROP_TAG(PT_NULL, tag);
 				}
 
-				auto names = GetNamesFromIDs(lpMAPIProp, &lpTag, NULL);
+				const HRESULT hRes = WC_H(GetNamesFromIDs(lpMAPIProp, lpTag, NULL, names));
 				MAPIFreeBuffer(lpTag);
-				return names;
+				return hRes;
 			}
 
-			return {};
+			return E_OUTOFMEMORY;
 		}
 
-		_Check_return_ std::vector<std::shared_ptr<namedPropCacheEntry>> GetAllNamesFromIDs(_In_ LPMAPIPROP lpMAPIProp)
+		_Check_return_ std::vector<std::shared_ptr<namedPropCacheEntry>>
+		GetAllNamesFromIDs(_In_ LPMAPIFOLDER lpMAPIFolder)
 		{
 			auto names = std::vector<std::shared_ptr<namedPropCacheEntry>>{};
 
 			// We didn't get any names - try manual
-			const auto ulLowerBound = __LOWERBOUND;
-			const auto ulUpperBound = FindHighestNamedProp(lpMAPIProp);
-			const ULONG batchSize = registry::namedPropBatchSize;
+			constexpr auto ulLowerBound = __LOWERBOUND;
+			const auto ulUpperBound = FindHighestNamedProp(lpMAPIFolder);
+			ULONG batchSize = registry::namedPropBatchSize;
 
 			output::DebugPrint(
 				output::dbgLevel::NamedProp,
 				L"GetAllNamesFromIDs: Walking through all IDs from 0x%X to 0x%X, looking for mappings to names\n",
 				ulLowerBound,
 				ulUpperBound);
-			for (auto iTag = ulLowerBound; iTag <= ulUpperBound && iTag < __UPPERBOUND; iTag += batchSize - 1)
+
+			auto iTag = ulLowerBound;
+			while (iTag <= ulUpperBound && iTag < __UPPERBOUND)
 			{
-				const auto range = GetRange(lpMAPIProp, iTag, min(iTag + batchSize - 1, ulUpperBound));
+				auto end = min(iTag + batchSize - 1, ulUpperBound);
+				std::vector<std::shared_ptr<namedPropCacheEntry>> range;
+				range.reserve(batchSize);
+
+				// Trying to get a range of these props can fail with MAPI_E_CALL_FAILED if it's too big for the buffer
+				// In this scenario, reopen the object, lower the batch size, and try again
+				HRESULT hRes = WC_H(GetRange(lpMAPIFolder, iTag, end, range));
+				if (hRes == MAPI_E_CALL_FAILED)
+				{
+					LPMAPIFOLDER newFolder = nullptr;
+					hRes = WC_H(GetRootFolder(lpMAPIFolder, &newFolder));
+
+					// If we can't re-open the root folder, something is bad - just break out
+					if (FAILED(hRes))
+					{
+						break;
+					}
+
+					// Swap the folder, lower the batch size, and try again
+					lpMAPIFolder->Release();
+					lpMAPIFolder = newFolder;
+					batchSize /= 2;
+					continue;
+				}
+				else if (FAILED(hRes))
+				{
+					// Other errors are unexpected
+					break;
+				}
+
 				for (const auto& name : range)
 				{
 					if (name->getMapiNameId()->lpguid != nullptr)
 					{
 						names.push_back(name);
 					}
 				}
+
+				// Go to the next batch
+				iTag += batchSize - 1;
```

Comment:
I know this came from my loop, but I think this should just be += batchsize. We're double searching at the boundaries, and batchsize of 1 will never progress.

Reply:
Good catch.